File: app/src/main/java/workshop1024/com/xproject/view/floatbutton/FloatingActionButtonBehavior.java
package workshop1024.com.xproject.view.floatbutton;

import android.content.Context;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

//https://www.jianshu.com/p/c174edcce58d
public class FloatingActionButtonBehavior extends FloatingActionButton.Behavior {

    public FloatingActionButtonBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    @Override
    public boolean layoutDependsOn(CoordinatorLayout parent, FloatingActionButton child, View dependency) {
        return dependency instanceof AppBarLayout;
    }

    @Override
    public boolean onDependentViewChanged(CoordinatorLayout parent, FloatingActionButton child, View dependency) {
        Log.i("XProject", "dependency.getTop() = " + dependency.getTop());
        Log.i("XProject", "dependency.getHeight() = " + dependency.getHeight());
        //计算菜单栏隐藏了高度的百分比
        float translationPercent = (float) dependency.getTop() / dependency.getHeight();
        Log.i("XProject", "translationPercent = " + translationPercent);
        Log.i("XProject", "child.getHeight() = " + child.getHeight());
        Log.i("XProject", "child.getBottom() = " + child.getBottom());
        //使用该百分比和FloatAtionButton高度和底部边距相乘，计算FloatAtionButton应该移动的距离
        float translationY = Math.abs((child.getHeight() + (parent.getHeight() - child.getBottom())) * translationPercent);
        Log.i("XProject", "translationY = " + translationY);
        child.setTranslationY(translationY);
        return true;
    }
}

File: home/src/main/java/workshop1024/com/xproject/home/view/floatbutton/FloatingActionButtonBehavior.kt
package workshop1024.com.xproject.home.view.floatbutton

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import androidx.coordinatorlayout.widget.CoordinatorLayout
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton
import kotlin.math.abs

//https://www.jianshu.com/p/c174edcce58d
class FloatingActionButtonBehavior(context: Context, attrs: AttributeSet) : FloatingActionButton.Behavior(context, attrs) {

    override fun layoutDependsOn(parent: CoordinatorLayout, child: FloatingActionButton, dependency: View): Boolean {
        return dependency is AppBarLayout
    }

    override fun onDependentViewChanged(parent: CoordinatorLayout, child: FloatingActionButton, dependency: View): Boolean {
        Log.i("XProject", "dependency.getTop() =  ${dependency.top}")
        Log.i("XProject", "dependency.getHeight() = ${dependency.height}")

        //计算菜单栏隐藏了高度的百分比
        val translationPercent = dependency.top.toFloat() / dependency.height

        Log.i("XProject", "translationPercent = $translationPercent")
        Log.i("XProject", "child.getHeight() = ${child.height}")
        Log.i("XProject", "child.getBottom() = ${child.bottom}")

        //使用该百分比和FloatAtionButton高度和底部边距相乘，计算FloatAtionButton应该移动的距离
        val translationY = abs((child.height + (parent.height + child.bottom)) * translationPercent)
        Log.i("XProject", "translationY = $translationY")
        child.translationY = translationY
        return true
    }
}